### PR TITLE
New Xic0 -> eXi codes

### PR DESCRIPTION
**Pull request by Chongk**
Newly written Xic0 -> eXI analysis codes for Lego train run

### CODE REVIEW

**Comment by lvermunt** [comment]
Hi @Chongk, thanks! I will have a look later today. But first of all: you should include your new task also in https://github.com/alisw/AliPhysics/blob/master/PWGHF/vertexingHF/PWGHFvertexingHFLinkDef.h and https://github.com/alisw/AliPhysics/blob/master/PWGHF/vertexingHF/CMakeLists.txt

**Comment by lvermunt** [comment]
No, you can just commit and push to the master branch of your fork. It will then automatically be added to this PR

**Comment by Chongk** [comment]
Thank you. I'll do it. Just a moment...

**Comment by lvermunt** [comment]
Hi @ChongK, I had a first look and I have some doubts. Did you have a look at the documentation how to write an analysis task: https://alice-doc.github.io/alice-analysis-tutorial/analysis/welcome.html ? I'm not an expert on this, but I assume there will be issues when you try to run your current code on LEGO trains (maybe even when building the tag). Please have a look at the various subsections in the link above and act accordingly.

In addition:
- Could you stay with the AliAnalysisTaskSE* convention (instead of AliAnaTaskSE*)?
- I see you are using a .C file instead of the *.cxx implementation file. I would suggest you to also stay here with the usual convention of a .h header file, a .cxx implementation file, and a .C addtask macro.
- Please copy the typical copyright lines in your .cxx class
- Please place your task among the other AnalysisTasks in CMakeList and PWGHFvertexingHFLinkDef instead of on the top.

More comments on the HF part later.

**Comment by Chongk** [comment]
Hello @lvermunt,

Guess I have to withdraw this PR and commit again. I'll keep this open until you finish all comments.
Also, it's already 8 PM here - I'll send the next ping tomorrow (at least 12 hours later).

By the way, for your previous comments,

1. Could you be more specific about what's problematic?
For the link you posted, yes, I have been referring to that when I started the analysis.
I've thought if the task works on Grid environment then there's no problem it runs on Lego train, either.
But judging from your worry it does NOT? If so, could you let me know which point made you worry?

2. I'll change those conventions (AliAnalysisTaskSE, *.cxx, etc) next commit.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 1]
```diff
@@ -0,0 +1,947 @@
...
+		fEvtPileupMB   = fEvtCutsMB  ->IsEventRejectedDueToPileup();
+		fEvtPileupHMV0 = fEvtCutsHMV0->IsEventRejectedDueToPileup();
+		if (fEvtPileupMB && fEvtPileupHMV0) return; //Return 8d
+		else fHisto->FillTH1("EvtCut", "PileupX>0", 1); //@
+
+		//Check vtx validity (NOT vtxZ)
+		fEvtVtx = fEvt->GetPrimaryVertex();
+		if (!fEvtVtx || fEvtVtx->GetNContributors()<cut_vtxNContributors) return; //Return 9d
+
+		//Store AliNormalizationCounter: DO NOT CHANGE THIS POSITION! (after vtx found, before |VtxZ| < 10)
```
Comment (lvermunt) [inline PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:230]: Are you sure this is the correct position? Please check wrt other tasks. It should be done before the RDHFCuts->IsEventSelected() call. But since you don't use that in the same way, please make sure it is at the same position

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 2]
```diff
@@ -0,0 +1,947 @@
...
+		fEvtVtx = fMCEvt->GetPrimaryVertex();
+		fEvtVtxZ = fEvtVtx->GetZ();
+		if (fabs(fEvtVtxZ) > cut_vtxZ) return; //Return 7mc
+
+		//Get true Xic0
+		int nXic0 = 0;
+		const int nTracksMC = fMCEvt->GetNumberOfTracks();
+		for (int a=0; a<nTracksMC; a++)
+		{
+			fMCPart = (AliAODMCParticle*)fMCEvt->GetTrack(a);
```
Comment (lvermunt) [inline PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:278]: This is not the same as storing all generated Xic (i.e. incorrect for the efficiency and acceptance correction). You need to loop over the MC generated TClonesArray instead. See for example: https://github.com/alisw/AliPhysics/blob/master/PWGHF/vertexingHF/vHFML/AliAnalysisTaskSEHFTreeCreatorApply.cxx#L1710

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 3]
```diff
@@ -0,0 +1,947 @@
...
+			if (!fMCPart || abs(fMCPart->GetPdgCode())!=PDGCode_Xic0) continue; //Continue 1
+
+			//Search e-Xi pair among daughters
+			bool t_found_e = false;
+			bool t_found_Xi = false;
+			AliAODMCParticle* fMCPart_e = 0;
+			AliAODMCParticle* fMCPart_Xi = 0;
+			for (int b=fMCPart->GetDaughterFirst(); b<=fMCPart->GetDaughterLast(); b++)
```
Comment (lvermunt) [inline PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:286]: You need to check if the numbers of daughters is the number you want. Right now you also select the four-body decays

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 2 above)
Comment (lvermunt) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:278]: It is also not the same as storing reconstructed Xic (but I don't think that was your aim here). For the proper efficiency correction you need to store both the generated (as above) and the reconstructed+selected eXi pairs that are matched to a Xic at MC truth level

**Review by lvermunt — COMMENTED** [review]
Hi @Chongk, see my review for some comments. The main ones are regarding:
 - You are storing the generated Xic in the wrong way. Also I seriously doubt if you are able to tag reconstructed + selected Xic candidates (in the correct decay chain) with your current structure with separated TTrees for Cascades and electrons. How do you plan to do this? You need it for the efficiency correction.
 - I think you have non-unique event IDs, which will lead to matching electrons with cascades from different events
 - Freedom to vary selections
 
On your questions: 
 - I was doubting specifically the constructors and streamers. But I'm not expert. So I was just suggesting to cross check with the details there. In general, that a precompiled task works in your local GRID environment doesn't mean that it also works with tags via the LEGO train environment.
 - As you prefer. Or you close and open a new one if you are ready, or you make this a draft PR

**Comment by Chongk** [comment]
Hi @lvermunt, thanks for your time.

I'll address your comments above by replying or by another mail - I assume it'll take some time.
I withdraw this PR and will request it again once I am ready.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 4]
```diff
@@ -0,0 +1,947 @@
...
+
+#include <cmath>
+#include <iostream>
+#include <vector>
+using namespace std;
+
+AliAnaTaskSEXic0SL::AliAnaTaskSEXic0SL():AliAnalysisTaskSE("AliAnaTaskSEXic0SL")
+{
+	ControlAnaObjects(0);
+	ControlOutputContainers(0);
```
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:46]: Done

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 5]
```diff
@@ -0,0 +1,947 @@
...
+	return true;
+}//FilterCascade
+
+//=======================================================================================
+
+void AliAnaTaskSEXic0SL::ControlOutputContainers(int option)
+{
+	if (option == 0)
+	{
+		DefineOutput(1, TDirectory::Class());
```
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:657]: Done

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 6]
```diff
@@ -0,0 +1,947 @@
...
+	//-----------------------------------------------------
+
+	//Check AOD event object
+	AliVEvent *event = InputEvent();
+	if (event->IsA() == AliAODEvent::Class()) fEvt = dynamic_cast<AliAODEvent*>(event);
+	else { Printf("ERROR: input event type is NOT AOD"); return; } //Return 1
+
+	//Check run number
+	fEvtRunNo = fEvt->GetRunNumber();
+	if (fEvtRunNo<cut_runNoLo || fEvtRunNo>cut_runNoUp) { Printf("ERROR: runNo"); return; } //Return 2
```
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:172]: Done. Please note that those cut values still exist in the header. They'll be used later offline for the sanity checks.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 7]
```diff
@@ -0,0 +1,947 @@
...
+	fANC_INEL0_HMV0_0to0p1->SetStudyMultiplicity(true, 1.); fANC_INEL0_HMV0_0to0p1->Init();
+
+	ControlOutputContainers(1);
+	return;
+}//UserCreateOutputObjects
+
+//=======================================================================================
+void AliAnaTaskSEXic0SL::UserExec(Option_t *)
+{
+	fEvtID++;
```
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:159]: That event ID was only used for the comparison between old vs. new codes, by using an AOD segment as a sample.

It's doubtful to me event by event identification would matter much in later analysis steps -
I guess the run number would be enough.

[the remark this answers is not in the document]
Comment (lvermunt) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:159]: I see now indeed you only have one TTree, with an `Event`, `Casc`, and `Ele` branch. When I made the comment I thought you had three independent TTrees, for which it is for sure required.

If in this case you would need a unique eventID still depends on how you will process this TTree offline. Do you plan to read it on event-by-event basis? Or separately load the `Event`, `Casc`, and `Ele` branches in three objects (as in a dataframe way)? In the latter case I think you need a unique eventID to not match cascades and electrons belonging to different collisions.

To prepare the mixed-event distributions, just using a different `runnumber` is not be enough. It is important to only mix events that have a similar z-vertex position and similar multiplicity values. Are you able to do this without a unique eventID?

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 8]
```diff
@@ -0,0 +1,947 @@
...
+	fHisto->CreateTH1("c_massXi", "", 400, 1.1, 1.5, "s");
+	fHisto->CreateTH1("c_massXi_MB", "", 400, 1.1, 1.5, "s");
+	fHisto->CreateTH1("c_massXi_HMV0", "", 400, 1.1, 1.5, "s");
+
+	//Tree
+	fTree = new TTree("T", Form("Xic0SemiLeptonic%s%s", IsPA?"_pA":"", IsMC?"_MC":""));
+	ControlOutputTree(fTree, IsMC);
+
+	//AliNormalizationCounter
+	fANC_MB_0to100   = new AliNormalizationCounter("ANC_MB_0to100");
```
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:134]: The arguments of the function you informed are integers. For HMV0 and MB + [0.1, 30], I cannot use that function.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 7 above)
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:159]: 1.
The basic strategy is processing them on an event-by-event basis, the former case in your explanation.

2.
For event mixing, I didn't think about the necessity of putting multiple mixing pools by the zVtx as you explained.

We're still working on writing the mixing code (actually it's not even started yet),
we planned to prepare a "buffer" structure to store an event temporarily.

I guess the concept itself has no problem (and Sanghoon already has experience with it),
and if the sort of events out by the zVtx matters having multiple buffers would be enough.
Well, easier said than done... I'll consider assigning a unique ID again before I push this code.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 9]
```diff
@@ -0,0 +1,947 @@
...
+		const bool t_trigFiredMB   = (fEvtTrig & AliVEvent::kINT7)?true:false;
+		const bool t_trigFiredHMV0 = (fEvtTrig & AliVEvent::kHighMultV0)?true:false;
+
+		fHisto->FillTH1("EvtTrig", "Any", 1); //@
+		if (t_trigFiredMB) fHisto->FillTH1("EvtTrig", "MB", 1); //@
+		if (t_trigFiredHMV0) fHisto->FillTH1("EvtTrig", "HMV0", 1); //@
+
+		//Activate RDHF cut by invoking it, then check pileup status by using multi vertexer
+		const bool t_evtGoodMB   = fEvtCutsMB  ->IsEventSelected(fEvt);
+		const bool t_evtGoodHMV0 = fEvtCutsHMV0->IsEventSelected(fEvt);
```
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:220]: I'm a little confused. If you're asking "capability of check which trigger fired",
I save the trigger bit itself ( AliInputEventHandler->IsEventSelected() ) and am able to check it offline.
Would you check the variable " fEvtTrig " ?

But if your intention is saving those two booleans from " fEvtCuts<> ", yes, I need to modify the code.
Is that what you meant?

[the remark this answers is not in the document]
Comment (lvermunt) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:220]: I mean that when analysing 0-0.1% you should (be able to) process **only** events that have the HMV0 trigger enabled (this includes HMV0 & kINT7, or HMV0 & any other trigger). And when you analyse the 0.1-30% and 30-100% event classes, you should process **only** events that have the INT7 (or INT7 & X) trigger. If you are able to do this with `fEvtTrig` fine.

[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:220]: For the capability issue, yes, I can do it offline.
For the "leaking" issue,

1. Case of look into " HMV0 + [0, 0.1] multiplicity ":
 a. Both HMV0 and kINT7 fired: ok
 b. Only HMV0 fired: ok (but this should be impossible)
 c. Only kINT7 fired: NOT ok
 
2. Case of look into "MB + [x, y]":
 a. Both kINT7 and HMV0 fired: ok
 b. Only kINT7 fired: ok
 c. Only HMV0 fired: NOT ok
 
 Is this what you mean? If so, then I do these judgments offline.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 10]
```diff
@@ -0,0 +1,947 @@
...
+	fEvtCutsMB->SetTriggerClass("");
+	fEvtCutsMB->SetTriggerMask(AliVEvent::kINT7);
+	fEvtCutsMB->SetUseInt7TriggerPP2012();
+	fEvtCutsMB->SetUsePhysicsSelection(true);
+
+	//Eventwise cut, for HMV0 trigger
+	fEvtCutsHMV0 = new AliRDHFCutsXictoeleXifromAODtracks();
+	fEvtCutsHMV0->SetOptPileup(AliRDHFCuts::kRejectMVPileupEvent);
+	fEvtCutsHMV0->SetTriggerClass("");
+	fEvtCutsHMV0->SetTriggerMask(AliVEvent::kHighMultV0);
```
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:84]: Done. I created another RDFH object and exclusively (with the existing ones) use it

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 11]
```diff
@@ -0,0 +1,947 @@
...
+
+	//AliESDTrackCut for track filtering
+	fTrkCuts = new AliESDtrackCuts();
+	fTrkCuts->SetClusterRequirementITS(AliESDtrackCuts::kSPD, AliESDtrackCuts::kBoth);
+	fTrkCuts->SetDCAToVertex2D(true);
+	fTrkCuts->SetMaxChi2PerClusterITS(cut_maxChi2PerClusterITS);
+	fTrkCuts->SetMaxDCAToVertexXY(cut_maxDCAToVertexXY);
+	fTrkCuts->SetMaxDCAToVertexZ(cut_maxDCAToVertexZ);
+	fTrkCuts->SetMinNClustersITS(cut_minNClustersITS); //ITS cluster
+	//fTrkCuts->SetMinNClustersTPC(fNClustersTPCMin = 70); //TPC min # of clusters - why masked?
```
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:95]: 1. I'll delete those parts to avoid bias

2. Two cuts you informed me of are being applied during offline selection.
Since the minimum cut values in the offline selection are " # of crossed rows: 65, Ratio: 0.7 ",
I'll set those values as 65 and 0.7.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C [hunk 12]
```diff
@@ -0,0 +1,947 @@
...
+
+	//Eventwise cut, for HMV0 trigger
+	fEvtCutsHMV0 = new AliRDHFCutsXictoeleXifromAODtracks();
+	fEvtCutsHMV0->SetOptPileup(AliRDHFCuts::kRejectMVPileupEvent);
+	fEvtCutsHMV0->SetTriggerClass("");
+	fEvtCutsHMV0->SetTriggerMask(AliVEvent::kHighMultV0);
+	fEvtCutsHMV0->SetUsePhysicsSelection(true);
+
+	//AliESDTrackCut for track filtering
+	fTrkCuts = new AliESDtrackCuts();
```
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:88]: I prepared an option in UserCreateOutputObjects, which set both fEvtCuts and TrkCuts by using an external file.
(* only an "if" statement for now, still need to implement it)

**Comment by lvermunt** [comment]
Hi @ChongK,

Thanks a lot for the work. Some replies:
 - Sure, fine to keep the old structure wrapped. This is indeed good for cross checks
 - ROOT file-based cut adjustments: see basically any D-meson task. For example `AliAnalysisTaskSED0Mass.cxx/h` + `AddTaskD0Mass.C`
 - Fine if you apply `MinNCrossedRows` and `MinRatioCrossedRowsOverFindableClusters` offline. However, please make sure you do apply the same selection for reco level in the efficiency calculation.
 - Normalization counter: sure, it is no problem to keep the separated ones. I wasn't aware it wouldn't work for V0M (also I'm a bit surprised, because with this task we actually did vs-V0M percentile analyses using the NormalizationCounter: `AliAnalysisTaskSEDvsMultiplicity.cxx`)
 - Event mixing: if you do the mixing online you probably don't need the eventID indeed. And yes, we typically use multiple buffers (usually we call them pools) for each z-vertex + multiplicity bin. You could have a look at the D0 mixed-event task for the definition see below (and elsewhere in the task for the use) https://github.com/alisw/AliPhysics/blob/6015b235c21bc8b9d00af9a764be1fb58f7bb32d/PWGHF/vertexingHF/AliAnalysisTaskCombinHF.cxx#L536-L557
 - I don't know is `fInputHandler->IsEventSelected()` is outdated. It might do the same as the `AliRDHFCuts::IsEventSelected()` call. Anyhow I would suggest you to use the latter one, because i) for the correct placing of the NormalizationCounters and ii) to not miss HF-specific event selections.
 - "Leaking" issue: Ok if you can do this offline. There is a point regarding the efficiency correction though, see below.

Also a followup on the efficiency correction:
 - Because you linked Jinjoo's task, I noticed I overlooked something. At first I thought you were looping over reconstructed tracks and matching those to MC truth ones (which is wrong, hence my comment). However, I see now you use the `AliMCEvent::GetNumberOfTracks()` and `AliMCEvent::GetTrack(a)` features. This might do the same as our usual strategy. If you can confirm this, you can also stay with what you do now.
 - The check on the number of daughters is still important to add. Could you check if this gives a bias in Jinjoo's analysis?
 - I repeat my question on how you plan to store the reconstructed electrons and cascades that are matched to (the same) MC true Xic (the numerator in the efficiency calculation). I believe this is a big missing point in your task.
 - Please have a look at section 4.4 of https://alice-notes.web.cern.ch/system/files/notes/analysis/993/2021-12-13-DLcvsmult_finalversion.pdf for the MC multiplicity reweighting. I think you are currently not able to do such checks. I believe they are required also in case you decide to use the minimum-bias efficiency (as we discussed via email) since then you need to prove the efficiency is stable with multiplicity.
 - Also keep in mind you need to be able to modify the generated MC pT shape of Xic as systematic variation

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 1 above)
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:230]: This is the function I use later, to retrieve # of events for normalization [link](http://alidoc.cern.ch/AliPhysics/master/_ali_normalization_counter_8cxx_source.html#l00179).

```
Double_t AliNormalizationCounter::GetNEventsForNorm()
{
    Double_t noVtxzGT10=GetSum("noPrimaryV")*GetSum("zvtxGT10")/GetSum("PrimaryV");
    return GetSum("countForNorm")-noVtxzGT10;
}
```

It applies correction for the case "|vtx_z|| > 10, for vertex object not found".
To do that, it estimates the ratio of "vtx object found"  and "vtx object found + |vtx_z| < 10".
If I |vtx_z| < 10 beforehand the estimation screws up, since there will be no "|vtx_z| > 10" yields at all.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 2 above)
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:278]: 1. The functions I use are:

a. AliMCEvent->GetNumberOfTracks() ([link] http://alidoc.cern.ch/AliRoot/v5-09-24/_ali_m_c_event_8h_source.html#l00098)) 
`virtual Int_t GetNumberOfTracks() const {return fNparticles;}`

If you check the function AliMCEvent::SetParticleArray(TClonesArray* mcParticles) ([link](http://alidoc.cern.ch/AliRoot/v5-09-24/_ali_m_c_event_8cxx_source.htm )),
```
fMCParticles = mcParticles;
fNparticles = fMCParticles->GetEntries();
```

b. AliMCEvent->GetTrack();
`virtual AliVParticle *GetTrack(Int_t i) const;

2. The functions in your reference are:

```
 for(Int_t iPart=0; iPart<arrayMC->GetEntriesFast(); iPart++){
    
    AliAODMCParticle* mcPart = dynamic_cast<AliAODMCParticle*>(arrayMC->At(iPart));
```

In conclusion, they look basically the same.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 3 above)
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:286]: 1. I checked by comparing "manual counting over loop <-> AliAODMCParticle->GetNDaughters()".
They're always the same. I'll update it to use GetNDaughters(), as it's more simple.

2. I added a boolean flag if either pi0 or charged pi is found among daughters.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 1 above)
Comment (lvermunt) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:230]: Since the `StoreEvent()` function does its [own event selection](https://github.com/alisw/AliPhysics/blob/master/PWGHF/vertexingHF/AliNormalizationCounter.cxx#L179) (based on the `AliRDHFCuts` object provided as argument [*]) and fills the relevant histograms based on the outcome of the various event selections. For this to properly work (i.e. you don't reject events in your analysis task that the `AliNormalizationCounter::StoreEvent()` function needs to access for the proper normalisation) it is best to put the call on top before any event selection. There are a few exceptions which aren't checked in the `AliRDHFCuts::IsEventSelected()` call; the placing of the `StoreEvent()` call should therefor be after:
 - the deltaAOD protection ([AliAnalysisTaskSEDs.cxx#L529-L541](https://github.com/alisw/AliPhysics/blob/master/PWGHF/vertexingHF/vHFML/AliAnalysisTaskSEDs.cxx#L529-L541)). Probably not relevant for your task though.
 - checking AOD pointer and TClonesArrays ([AliAnalysisTaskSEDs.cxx#L564-L568](https://github.com/alisw/AliPhysics/blob/master/PWGHF/vertexingHF/vHFML/AliAnalysisTaskSEDs.cxx#L564-L568))
 - checking for null vertex pointer and a non-zero magnetic field ([AliAnalysisTaskSEDs.cxx#L570-L572](https://github.com/alisw/AliPhysics/blob/master/PWGHF/vertexingHF/vHFML/AliAnalysisTaskSEDs.cxx#L570-L572))

In particular, I was doubting your previous placing due to the `IsEventRejectedDueToPileup()` and `GetNContributors()` selections that you placed above the `StoreEvent()` call.

Btw, please refer to https://github.com/alisw/AliPhysics for the latest code, http://alidoc.cern.ch/ is not up to date.

[*] This is another reason why I recommend you to use the IsEventSelected() from the AliRDHFCuts object instead of your own event-selection implementation, to avoid possible mismatches between both event selections leading to the wrong number of estimated events.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 3 above)
Comment (lvermunt) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:286]: (I don't fully follow your 1) and 2), please correct me if I understood wrongly)

1) It might be the case that for the MC you currently analysing all Xic have N=3 daughters, since the four-body Xic decays were not yet enabled when producing these MCs. However, as we are discussing via email, you need to request a new MC with additional Xic decays enabled for your analysis to do the background-subtraction template fit. And for that MC this will not hold anymore. So please add a condition on the correct number of Xic daughters when you tag Xic for the efficiency calculation.

2) Ok for calculating the reconstruction efficiency. I don't think however that labels of the truth Xic0 and e/Xi's mothers are enough to perform the background template fit. You also need to tag which four-body decay channel it is.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 10 above)
[the remark this answers is not in the document]
Comment (lvermunt) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:84]: Please write a new reply instead of editing it, so we get a notification. I noticed this by accident now.

1) Ok, fair point. But then please add a check if the event selections in both the MB and HM objects are the same. Something like: [AliAnalysisTaskSEHFTreeCreatorApply.cxx#L756-L761](https://github.com/alisw/AliPhysics/blob/master/PWGHF/vertexingHF/vHFML/AliAnalysisTaskSEHFTreeCreatorApply.cxx#L756-L761). Note you should change the trigger briefly to `kAny` for this check to work

2) Event selections are also important when analysing MC, please add the `AliRDHFCuts::IsEventSelected()` call also for MC. You should use the same objects as you give to data, just changing the trigger to `kAny`

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 11 above)
[the remark this answers is not in the document]
Comment (lvermunt) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:95]: Your edit of today made me think you missed my long reply of two days ago (where I said it was fine to apply them offline). Did you see https://github.com/alisw/AliPhysics/pull/21512#issuecomment-1358985604 ?

Again, please write a new reply instead of editing an existing one

[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:95]: "Make sure to apply the same cuts in reco level", yes, I read them through.
The edit today was just a notification for not adding those cuts in RDHF objects,
in case you check the updated code later and wonder why they're not added.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 3 above)
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:286]: Ok, now I got your point. I updated the truth MC handling as follows.

1. Possible 4-body decay modes are these two:
 a. Xic+ -> (e+) + (nu) + (Xi*0) -> (e+) + (nu) + (Xi- + pi+)
 b. Xic0 -> (e-) + (nu_bar) + (Xi*+) -> (e-) + (nu_bar) + (Xi+ + pi0)

2. At the truth level,
 a. Truth Xic+ also will be accepted in addition to the Xic0
 b. Added label and PDG of truth particle
 
The distinction should be possible If I compare 2-b with reco. electron's (or Xi) mother label/PDG.
For the case of 4-body decay only, requiring Xi's mother PDG to be either Xi*0 or Xi*+ should be sufficient.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 10 above)
[the remark this answers is not in the document]
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:84]: 1.
The following lines are the current code implementation, temporarily modified for the test you asked for.

```
        //Eventwise cut, for MB (default)
        fEvtCutsMB = new AliRDHFCutsXictoeleXifromAODtracks();
        fEvtCutsMB->SetOptPileup(AliRDHFCuts::kRejectMVPileupEvent); //Multi vertexer pileup rejection
        if (IsMC==false)
        {
            //fEvtCutsMB->SetTriggerMask(AliVEvent::kINT7);
            //fEvtCutsMB->SetUseInt7TriggerPP2012();
            fEvtCutsMB->SetTriggerMask(AliVEvent::kAny);
        }
        else if (IsLegacy==false) fEvtCutsMB->SetTriggerMask(AliVEvent::kAny);
        fEvtCutsMB->SetUsePhysicsSelection(true);

        //Eventwise cut, for HMV0 trigger
        fEvtCutsHMV0 = new AliRDHFCutsXictoeleXifromAODtracks();
        fEvtCutsHMV0->SetOptPileup(AliRDHFCuts::kRejectMVPileupEvent);
        if (IsMC==false)
        {
            //fEvtCutsHMV0->SetTriggerMask(AliVEvent::kHighMultV0);
            fEvtCutsHMV0->SetTriggerMask(AliVEvent::kAny);
        }
        else if (IsLegacy==false) fEvtCutsHMV0->SetTriggerMask(AliVEvent::kAny);
        fEvtCutsHMV0->SetUsePhysicsSelection(true);
```
And then, I tested if the response is the same by using a sample AOD, as follows.
```
        //Activate RDHF cut by invoking it
        fEvtGoodMB   = fEvtCutsMB  ->IsEventSelected(fEvt);
        fEvtGoodHMV0 = fEvtCutsHMV0->IsEventSelected(fEvt);

        cout <<Form("EvtID: %5i, IsEvtSelected - MB/HMV0: %i/%i %s\n",
                fEvtID, fEvtGoodMB, fEvtGoodHMV0, (fEvtGoodMB==fEvtGoodHMV0)?"":"!@#");
```
The results are exactly the same. I'll send the log if you need to check.

2.

File: PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C (hunk 1 above)
Comment (Chongk) [reply PWGHF/vertexingHF/AliAnaTaskSEXic0SL.C:230]: Done. For its updated position it'd be better to discuss it again in the next pull request.

**Comment by Chongk** [comment]
Hello @lvermunt,

I guess I addressed most of your comments above, except for a couple of items.
They're "eventID" and "loading cuts via an external ROOT file",
but I think they can be added later by necessity. These codes will surely need further updates.
For issues in your long post above, I believe most of them are already answered as replies. But for the following two,

1. Efficiency vs. Multiplicity:
I believe this is an issue already discussed between Sanghoon and Cristina.
If my memory serves, it's concluded we're not going to apply efficiency by multiplicity.

2. pT shape
Yes, I'm aware of it - there's already an offline analysis routine to address this.
Please have a look: [link](https://indico.cern.ch/event/1101855/contributions/4685342/attachments/2376650/4059989/Xic0_220121_CKim_D2H.pdf), especially for p9.

Let's continue this discussion after you're back from vacation.
Since I'm not sure if this discussion page will be intact if I issue another pull request,
please check the up-to-date codes in my private repository ([link](https://github.com/Chongk/Xic0New)). I already sent an invitation.

**Comment by lvermunt** [comment]
Hi @Chongk, thanks for addressing my comments. When you think you are ready, I would say you can open a new PR where we can continue the review.

**Comment by Chongk** [comment]
Hello @lvermunt,

Thanks for the ping. I'm doing some debugging now - I guess it'll take at least a day or two.
